### ISSUE TITLE
Question about homography estimation, multiplication of H, matrix M and its inverse

### ISSUE THREAD

**Issue by Melody-doudou** [nie-lang/UnsupervisedDeepImageStitching#10]
Thanks a lot for sharing your work!
I am having one question about the homography estimation part. I understand the process is that H is calculated by solving DLT and then used to warp the warped feature maps. I am not sure why H is multiplied by a matrix M and its inverse before warping the feature maps. What is the effect of matrix M? Related codes are as follows. Thank you! 

    H1 = solve_DLT(net1_f/4., patch_size)
    M = np.array([[patch_size / 2.0, 0., patch_size / 2.0],
                  [0., patch_size / 2.0, patch_size / 2.0],
                  [0., 0., 1.]]).astype(np.float32)
    M_tensor = tf.constant(M, tf.float32)
    M_tile = tf.tile(tf.expand_dims(M_tensor, [0]), [batch_size, 1, 1])
    M_inv = np.linalg.inv(M)
    M_tensor_inv = tf.constant(M_inv, tf.float32)
    M_tile_inv = tf.tile(tf.expand_dims(M_tensor_inv, [0]), [batch_size, 1, 1])
    H1 = tf.matmul(tf.matmul(M_tile_inv, H1), M_tile)
    feature2_warp = transform(tf.nn.l2_normalize(feature2[-2],axis=3), H1)

**Comment by nie-lang** [comment]
Well, in our warping process, we normalize the height [0, H] and width  [0, W] coordinates of images/features into [-1, 1]. Therefore, the corresponding homography should be normalized ( tf.matmul(tf.matmul(M_tile_inv, H1), M_tile) ) as well.

Of course, If you do not normalize the coordinates, the homography should be not normalized.

You can refer to Section ⅣC of https://arxiv.org/pdf/1709.03966.pdf for more details.

**Comment by Melody-doudou** [comment]
> Well, in our warping process, we normalize the height [0, H] and width [0, W] coordinates of images/features into [-1, 1]. Therefore, the corresponding homography should be normalized ( tf.matmul(tf.matmul(M_tile_inv, H1), M_tile) ) as well.
> 
> Of course, If you do not normalize the coordinates, the homography should be not normalized.
> 
> You can refer to Section ⅣC of https://arxiv.org/pdf/1709.03966.pdf for more details.

Oh okay, I see. Thank you very much! :)
Btw, the series of works about image stitching that you have done are really impressive!

**Comment by DanilHu** [comment]
Hi, I find another question about this part. I can get the point of using M and its inverse. But What's meaning of patch-size = 32? 
```python
# Input size is [128, 128] and feature[-1] should be [16*16] because of 3 times downsampling
def feature_extractor(image_tf): 
    feature = []
    #image_tf = tf.expand_dims(image_tf, [3])
    with tf.variable_scope('conv_block1'): # H
      conv1 = _conv_block(image_tf, ([64, 64]), (3, 3), (1, 1))
      feature.append(conv1)
      maxpool1 = slim.max_pool2d(conv1, 2, stride=2, padding = 'SAME')
    with tf.variable_scope('conv_block2'):
      conv2 = _conv_block(maxpool1, ([64, 64]), (3, 3), (1, 1))
      feature.append(conv2)
      maxpool2 = slim.max_pool2d(conv2, 2, stride=2, padding = 'SAME')
    with tf.variable_scope('conv_block3'):
      conv3 = _conv_block(maxpool2, ([128, 128]), (3, 3), (1, 1))
      feature.append(conv3)
      maxpool3 = slim.max_pool2d(conv3, 2, stride=2, padding = 'SAME')
    with tf.variable_scope('conv_block4'):
      conv4 = _conv_block(maxpool3, ([128, 128]), (3, 3), (1, 1))
      feature.append(conv4)
    
    return feature

    with tf.variable_scope('Reggression_Net1'):    
      search_range = 16
      global_correlation = cost_volume(tf.nn.l2_normalize(feature1[-1],axis=3), tf.nn.l2_normalize(feature2[-1],axis=3), search_range)   
      #3-convolution layers
      net1_conv1 = conv2d(inputs=global_correlation, num_outputs=512, kernel_size=3, activation_fn=tf.nn.relu)
      net1_conv2 = conv2d(inputs=net1_conv1, num_outputs=512, kernel_size=3, activation_fn=tf.nn.relu)
      net1_conv3 = conv2d(inputs=net1_conv2, num_outputs=512, kernel_size=3, activation_fn=tf.nn.relu)  
      # Flatten dropout_conv4
      net1_flat = slim.flatten(net1_conv3)
      # Two fully-connected layers
      with tf.variable_scope('net1_fc1'):
        net1_fc1 = slim.fully_connected(net1_flat, 1024, activation_fn=tf.nn.relu)
        net1_fc1 = slim.dropout(net1_fc1, keep_prob)
      with tf.variable_scope('net1_fc2'):
        net1_fc2 = slim.fully_connected(net1_fc1, 8, activation_fn=None) #BATCH_SIZE x 8
    
    net1_f = tf.expand_dims(net1_fc2, [2])
    # Why there is 32
    patch_size = 32.
    H1 = solve_DLT(net1_f/4., patch_size)
    M = np.array([[patch_size / 2.0, 0., patch_size / 2.0],
                  [0., patch_size / 2.0, patch_size / 2.0],
                  [0., 0., 1.]]).astype(np.float32)
    M_tensor = tf.constant(M, tf.float32)
    M_tile = tf.tile(tf.expand_dims(M_tensor, [0]), [batch_size, 1, 1])
    M_inv = np.linalg.inv(M)
    M_tensor_inv = tf.constant(M_inv, tf.float32)
    M_tile_inv = tf.tile(tf.expand_dims(M_tensor_inv, [0]), [batch_size, 1, 1])
    H1 = tf.matmul(tf.matmul(M_tile_inv, H1), M_tile)
    feature2_warp = transform(tf.nn.l2_normalize(feature2[-2],axis=3), H1)
```

At first I consider it as the size of feature map, but then I find the real size of the feature that used in global correlation is 16*16.
